nmrfx-core/src/main/java/org/nmrfx/chemistry/Polymer.java

```diff
@@ -373,24 +373,24 @@ public boolean isDNA() {
     public List<String> getCyclicConstraints() {
 
         List<String> constraints = new ArrayList<>();
-
-        for (int i = 0; i < cyclicClosers.length; i += 3) {
-            Atom atom1 = getLastResidue().getAtom(cyclicClosers[i]);
-            Atom atom2 = getFirstResidue().getAtom(cyclicClosers[i + 1]);
-            if (atom1 == null) {
-                System.out.println("no atom1 " + cyclicClosers[i]);
-                atom1 = getLastResidue().getAtom("H1");
-            }
-            if (atom2 == null) {
-                atom2 = getFirstResidue().getAtom("H1");
-            }
-            if (atom2 == null) {
-                System.out.println("no atom2 " + cyclicClosers[i + 1]);
+        if ((getLastResidue() != null) && (getFirstResidue() != null)) {
+            for (int i = 0; i < cyclicClosers.length; i += 3) {
+                Atom atom1 = getLastResidue().getAtom(cyclicClosers[i]);
+                Atom atom2 = getFirstResidue().getAtom(cyclicClosers[i + 1]);
+                if (atom1 == null) {
+                    System.out.println("no atom1 " + cyclicClosers[i]);
+                    atom1 = getLastResidue().getAtom("H1");
+                }
+                if (atom2 == null) {
+                    System.out.println("no atom2 " + cyclicClosers[i + 1]);
+                    atom2 = getFirstResidue().getAtom("H1");
+                }
+                if (atom1 != null && atom2 != null) {
+                    String distance = cyclicClosers[i + 2];
+                    String constraint = atom2.getFullName() + " " + atom1.getFullName() + " " + distance;
+                    constraints.add(constraint);
+                }
```

Comment:
Is it fine to just not add the constraint to the list being returned or should I raise an exception here? This method is called only once here: https://github.com/nanalysis/nmrfx/blob/4145ec117ba89f6926ff2be1dc000cdcaaa25900/nmrfx-structure/src/main/resources/refine.py#L792

Reply:
We should throw an exception.  If atom1 or atom2 is null at this point something was wrong with list of atoms to constrain.

Reply:
Used requireNonNull on the second attempt to set atom1 and atom2